examples/air-purifier-app/air-purifier-common/src/air-purifier-manager.cpp

```diff
@@ -191,13 +194,75 @@ void AirPurifierManager::HandleFanControlAttributeChange(AttributeId attributeId
     }
 }
 
+void AirPurifierManager::HandleOnOff(AttributeId attributeId, uint8_t type, uint16_t size, uint8_t * value)
+{
+    if (attributeId != OnOff::Attributes::OnOff::Id)
+    {
+        return;
+    }
+    bool on = static_cast<bool>(*value);
+    uint8_t new_speed;
+    uint8_t new_percent;
+    if (on)
+    {
+        ChipLogProgress(NotSpecified, "======================== Handle ON");
+        // If either of these come back as NULL, that should mean the fan is operating in auto mode.
+        // I have no idea what that means for this case, so I'll just set them to high because this
+        // is just an example.
+        // In theory these should always be NULL together or not at all, so hopefully
+        // the checks for only percent.IsNull() or only speed.IsNull() are more theoretical
+        // than practical.
+        DataModel::Nullable<Percent> percent = GetPercentSetting();
+        DataModel::Nullable<uint8_t> speed   = GetSpeedSetting();
+        uint8_t speedMax                     = 100;
+        FanControl::Attributes::SpeedMax::Get(mEndpointId, &speedMax);
+        if (percent.IsNull() && speed.IsNull())
+        {
+            // Operating in auto mode, set to 100
+            new_percent = 100;
+            new_speed   = speedMax;
+        }
+        else if (percent.IsNull())
+        {
+            // This should never happen, but nonetheless, let's check and warn.
+            ChipLogError(NotSpecified,
+                         "AirPurifierManager::HandleOnOff: PercentSetting is null when SpeedSetting is not. Setting PercentCurrent "
+                         "from Speed");
+            new_speed   = speed.Value();
+            new_percent = static_cast<uint8_t>(new_speed * 100u / speedMax);
+        }
+        else if (speed.IsNull())
+        {
+            // This should never happen, but nonetheless, let's check and warn.
+            ChipLogError(NotSpecified,
+                         "AirPurifierManager::HandleOnOff: SpeedSetting is null when PercentSetting is not. Setting SpeedCurrent "
+                         "from Percent");
+            new_percent = percent.Value();
+            new_speed   = static_cast<uint8_t>(new_percent * speedMax / 100u);
+        }
+        else
+        {
+            new_percent = percent.Value();
+            new_speed   = speed.Value();
+        }
+    }
+    else
+    {
+        ChipLogProgress(NotSpecified, "======================== Handle OFF");
+        new_percent = 0;
+        new_speed   = 0;
+    }
+    FanControl::Attributes::SpeedCurrent::Set(mEndpointId, new_speed);
+    FanControl::Attributes::PercentCurrent::Set(mEndpointId, new_percent);
+    mOnOffClusterOn = on;
+}
+
 void AirPurifierManager::PercentSettingWriteCallback(uint8_t aNewPercentSetting)
 {
-    if (aNewPercentSetting != percentCurrent)
+    ChipLogDetail(NotSpecified, "AirPurifierManager::PercentSettingWriteCallback: %d", aNewPercentSetting);
```

Comment:
static cast the aNewPercentSetting to int. Same for similar?

Reply:
done.